pkg/lib/infra/whatsapp/on_premises.go

```diff
@@ -222,29 +222,46 @@ func (c *OnPremisesClient) login(ctx context.Context) (*UserToken, error) {
 	}
 	req.SetBasicAuth(c.Credentials.Username, c.Credentials.Password)
 
+	whatsappAPIErr := &WhatsappAPIError{
+		APIType: config.WhatsappAPITypeOnPremises,
+	}
+
 	resp, err := c.HTTPClient.Do(req)
 	if err != nil {
-		return nil, err
+		return nil, errors.Join(err, whatsappAPIErr)
 	}
 	defer resp.Body.Close()
+	whatsappAPIErr.HTTPStatusCode = resp.StatusCode
+
+	dumpedResponse, err := nethttputil.DumpResponse(resp, true)
+	if err != nil {
+		return nil, errors.Join(err, whatsappAPIErr)
+	}
+	whatsappAPIErr.DumpedResponse = dumpedResponse
 
 	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
-		return nil, fmt.Errorf("whatsapp: unexpected response status %d", resp.StatusCode)
+		// It was observed that when status code is 401, the body is empty.
+		// Thus parse error should be ignored.
+		errResp, parseErr := c.tryParseErrorResponse(resp)
+		if parseErr == nil {
+			whatsappAPIErr.ParsedResponse = errResp
+		}
+		return nil, whatsappAPIErr
```

Comment:
I think `tryParseErrorResponse` will not return error if response body is empty? It still try to read the empty body an therefore get 0 bytes.

Reply:
Yes. In this particular case, since DumpResponse returns successfully, resp.Body is stored in memory. So a subsequent `io.ReadAll` should not fail. See https://cs.opensource.google/go/go/+/refs/tags/go1.23.4:src/net/http/httputil/dump.go;l=25

Therefore parseErr should always be nil.